### PR TITLE
Modify MVTLoader to output binary without going through geojson

### PR DESCRIPTION
**Pull request by felixpalmer**
This PR modified the MVTLoader such that when it outputs binary data, it no longer uses geoJSON as an intermediate representation. 

As a result on large tiles the parsing speed is increased by a factor of 2X in the average case, and sometimes as much as 3X - naturally depending on the data. As the code was often creating temporary objects needlessly, the new code also puts less pressure on garbage collection.

Care was taken to make sure the output data is equivalent to what the library currently does - this means that the bug noted in https://github.com/visgl/loaders.gl/issues/1274 is also present here.

A number of tests were added to check the validity of the new code as well as the speed. For these to run and to reproduce the results stated above, you'll need to download some test data, see the script in `modules/mvt/test/data/fetch_data.py`. Running this will fetch a selection of large MVT tiles, and print a list that can then be copied and pasted into the tests.

### CODE REVIEW

**Comment by felixpalmer** [comment]
The bulk of the work was done in `binary-vector-tile/features-to-binary.js` and `binary-vector-tile/vector-tile-feature.js`.

These follow a similar high level logic to the old code, except that the data is represented as a flat array from the start in `loadGeometry`. Thus there is an intermediate format like before, which is somewhat similar to geoJSON, but it is much closer to what the final format sent to the buffers ends up like.

`featuresToBinary` is then modified to take advantage of this and in many places there are optimizations compared to the old code, often in the form of for loops being removed and instead `TypedArray.set` or `TypedArray.fill` being used to copy large amounts of data quickly. The first pass over the data has also been removed and is now performed as part of the binary vector tile parsing.

File: modules/mvt/src/mvt-loader.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ export const MVTWorkerLoader = {
     'application/x-protobuf'
     // 'application/octet-stream'
   ],
-  worker: true,
+  worker: false,
```
Comment (ibgreen) [inline modules/mvt/src/mvt-loader.js:25]: This was just for testing? You can use `_workerType: 'test'` to build against local workers created by `yarn build`.

File: modules/mvt/test/mvt-loader.bench.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import {MVTLoader} from '@loaders.gl/mvt';
+import {setLoaderOptions, fetchFile, parse} from '@loaders.gl/core';
+import {geojsonToBinary} from '@loaders.gl/gis';
+
+setLoaderOptions({
+  _workerType: 'test'
```
Comment (ibgreen) [inline modules/mvt/test/mvt-loader.bench.js:6]: This sets it globally for all bench marks. Better to set it in each call.

File: modules/mvt/test/mvt-loader.binary.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import test from 'tape-promise/tape';
+import {MVTLoader} from '@loaders.gl/mvt';
+import {setLoaderOptions, fetchFile, parse} from '@loaders.gl/core';
+import {geojsonToBinary} from '@loaders.gl/gis';
+
+setLoaderOptions({
```
Comment (ibgreen) [inline modules/mvt/test/mvt-loader.binary.js:6]: Ditto

File: modules/mvt/test/data/fetch_data.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#!/usr/local/bin/python3
```
Comment (ibgreen) [inline modules/mvt/test/data/fetch_data.py:1]: Does this run on CI?

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.d.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import VectorTileFeature from './vector-tile-feature';
+
+export default class VectorTileLayer{
+  // Public
+  version: number;
```
Comment (ibgreen) [inline modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.d.ts:5]: `readonly version: number`?

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.d.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import Protobuf from 'pbf';
+
+export default class VectorTileFeature {
```
Comment (ibgreen) [inline modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.d.ts:3]: Nice to add short JSDoc on all classes and functions, in this case perhaps stating that this is mapbox pbf wrapper or something. It shows up in vscode when hovering, helps get oriented.

File: modules/mvt/src/lib/binary-vector-tile/features-to-binary.js [hunk 7]
```diff
@@ -0,0 +1,314 @@
+// Convert binary features to flat binary arrays
+// @ts-nocheck
+
+export function featuresToBinary(features, firstPassData, options = {}) {
```
Comment (ibgreen) [inline modules/mvt/src/lib/binary-vector-tile/features-to-binary.js:4]: Can we add JSDoc header to each function and remove the ts-nocheck?

Small time investment that will help review and maintenance.

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.js [hunk 8]
```diff
@@ -0,0 +1,211 @@
+// This code is forked from https://github.com/mapbox/vector-tile-js under BSD 3-clause license.
```
Comment (ibgreen) [inline modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.js:1]: Nit: I would put the code that is significantly unique to us in a different folder from the more or less completely forked code.

File: modules/mvt/test/mvt-loader.binary.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  lines: 'lines_2-2-1.mvt',
+  linesMultiple: 'lines_10-501-386_multiplelayers.mvt',
+  points: 'points_4-2-6.mvt',
+  polygons: 'polygons_10-133-325.mvt'
+};
+
+// Test to sanity check that old method of parsing binary
+// format via an intermediate geojson step produces the
+// same result
+for (const [name, filename] of Object.entries(testfiles)) {
```
Comment (ibgreen) [inline modules/mvt/test/mvt-loader.binary.js:22]: +1

**Review by ibgreen — APPROVED** [review]
@felixpalmer very solid work! 

I am OK with with merging this into master as is, but I do think we should do some additional cleanup PRs - initial round of nits in the code.

As we get this in place, we could take another look at your intermediate representation and see if there is more that can be done perf-wise.

Another improvement could be to also offer to do tesselation in the worker. The first step would be to make a binary version of `earcut` probably in math.gl and optionally call it here.

Slightly longer term, I'd love to generalize some of the binary generation utilities, specifically to see if we could benefit from them in the geojson etc loaders.

We have to decide whether to backport this to 2.3 or ask users to specify `3.0.0-alpha` until we release.

**Comment by ibgreen** [comment]
Please update the docs/whats-new section with the good news.

**Comment by alasarr** [comment]
Great job @felixpalmer !

@felixpalmer very solid work!

In relation to worker tesselation, I talked with @felixpalmer and @Pessimistress and it will be the next one to add. 

> Slightly longer term, I'd love to generalize some of the binary generation utilities, specifically to see if we could benefit from them in the geojson etc loaders.

I included some utilities in deck.gl at [https://github.com/visgl/deck.gl/blob/master/modules/layers/src/geojson-layer/geojson-binary.js](geojson-binary.js) and I agree it could be useful. Should it be part of a new utils module?  Or just a submodule of `loaders.gl`? 

> We have to decide whether to backport this to 2.3 or ask users to specify 3.0.0-alpha until we release.
Since deck.gl 8.4 has the binary property to false by default at MVTLayer, I think would be great to include in 2.3, test on 8.4, and switch on 8.5 to binary by default 💪

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.d.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  properties: object;
+  extent: any;
+  type: number;
+  id?: number;
+
+	constructor(pbf: Protobuf, end, extent, keys, values, firstPassData);
+
+	loadGeometry(); any;
+	classifyRings([number]): [[number]];
+  toLines(options: {x: number, y: number, z: number} | (([number], VectorTileFeature) => void)): object;
```
Comment (kylebarron) [inline modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.d.ts:15]: Nit: remove mixed tabs and spaces.

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.js [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+ */
+function classifyRings(rings) {
+  return rings.map(r => [r]);
+}
+
+function readFeature(tag, feature, pbf) {
+  if (tag === 1) feature.id = pbf.readVarint();
+  else if (tag === 2) readTag(pbf, feature);
+  else if (tag === 3) feature.type = pbf.readVarint();
+  else if (tag === 4) feature._geometry = pbf.pos;
```
Comment (kylebarron) [inline modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.js:200]: Nit: use a switch here?

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
+// This code is forked from https://github.com/mapbox/vector-tile-js under BSD 3-clause license.
+
+import VectorTileFeature from './vector-tile-feature.js';
+
+export default class VectorTileLayer {
+  constructor(pbf, end) {
+    // Public
+    this.version = 1;
+    this.name = null;
+    this.extent = 4096;
```
Comment (kylebarron) [inline modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js:10]: Is it expected for this to be hardcoded?

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+            ? pbf.readDouble()
+            : tag === 4
+              ? pbf.readVarint64()
+              : tag === 5
+                ? pbf.readVarint()
+                : tag === 6
+                  ? pbf.readSVarint()
+                  : tag === 7
+                    ? pbf.readBoolean()
+                    : null;
```
Comment (kylebarron) [inline modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js:75]: In my opinion the successive indentation makes things less clear, and I think something like below is much more readable. Also, it seems a little surprising to me that each iteration of the `while` loop overwrites `value` instead of appending to it. Maybe worth a comment?

```js
let value = null;
switch (tag) {
  case 1:
    value = pbf.readString();
    break;
  case 2:
    value = pbf.readFloat();
    break;
  case 3:
    value = pbf.readDouble();
    break;
  case 4:
    value = pbf.readVarint64();
    break;
  case 5:
    value = pbf.readVarint();
    break;
  case 6:
    value = pbf.readSVarint();
    break;
  case 7:
    value = pbf.readBoolean();
    break;
  default:
    value = null;
    break;
}
```

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+  }
+}
+
+function readLayer(tag, layer, pbf) {
+  if (tag === 15) layer.version = pbf.readVarint();
+  else if (tag === 1) layer.name = pbf.readString();
+  else if (tag === 5) layer.extent = pbf.readVarint();
+  else if (tag === 2) layer._features.push(pbf.pos);
+  else if (tag === 3) layer._keys.push(pbf.readString());
+  else if (tag === 4) layer._values.push(readValueMessage(pbf));
```
Comment (kylebarron) [inline modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js:50]: If you don't want to use a switch, I'd at least suggest that the order of tag matching is sorted? Seems more readable to check in sorted order. Unless you think the distribution of tag value matches this order, and this minimizes equality checking, in which case I'd suggest a comment to that effect.

File: modules/mvt/src/lib/parse-mvt.js [hunk 15]
```diff
@@ -88,6 +105,19 @@ function getDecodedFeature(feature, options = {}) {
   return decodedFeature;
 }
 
+function getDecodedFeatureBinary(feature, options = {}) {
+  const decodedFeature = feature.toLines(
```
Comment (kylebarron) [inline modules/mvt/src/lib/parse-mvt.js:109]: Maybe something like `feature.toBinaryCoordinates` would be a clearer name?

File: modules/mvt/test/data/fetch_data.py (hunk 4 above)
Comment (kylebarron) [reply modules/mvt/test/data/fetch_data.py:1]: Also nit: `#!/usr/bin/env python3` is more portable.

File: modules/mvt/test/data/fetch_data.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+TILES = [
+    {'z': 4, 'x': 4, 'y': 6},
+    {'z': 4, 'x': 3, 'y': 6},
+    {'z': 5, 'x': 8, 'y': 12},
+    {'z': 6, 'x': 18, 'y': 24},
+    {'z': 7, 'x': 35, 'y': 50},
+    {'z': 8, 'x': 75, 'y': 96}
+]
+for name, dataset in DATASETS.items():
```
Comment (kylebarron) [inline modules/mvt/test/data/fetch_data.py:29]: So basically this script is downloading 36 tiles? How big in total is this data? Is it too big to just store in this repo? Is the marginal benefit of each additional xyz index large? Otherwise we could store just 6 tiles locally; one tile for each dataset.

File: modules/mvt/test/data/fetch_data.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    # Get url pattern for tiles
+    url = f'{SERVER}/tileset?source={dataset}&format=tilejson'
+    r = requests.get(url)
```
Comment (kylebarron) [inline modules/mvt/test/data/fetch_data.py:32]: Nit: use params for the request
```py
params = {
    'source' dataset,
    'format': 'tilejson'
}
r = requests.get(f'{SERVER}/tileset', params=params)
```

File: test/browser.js [hunk 18]
```diff
@@ -35,7 +35,7 @@ test.onFailure(window.browserTestDriver_fail);
 window.__VERSION__ = require('../lerna.json').version;
 
 test('Browser tests', t => {
-  require('./modules');
```
Comment (kylebarron) [inline test/browser.js:38]: Why this change? `test/modules.js` are intended to be run on both Node and Browser tests. ... and it seems like a circular import?

**Review by kylebarron — COMMENTED** [review]
- I'm not very knowledgeable about tree-shaking, but would there be an easy way here to tell the bundler not to include the original loader if you were planning to use only the binary format? E.g. it seems like after this PR the loader would include two copies of `vector-tile-js`; the original one from Mapbox used for the non-binary loader and the forked copy for the binary loader? Maybe the sizes aren't a big enough concern to worry about this?
- It does seem like there's decent duplication with `geojsonToBinary` from `@loaders.gl/gis`. Might be worth thinking about how we can make that code more general to reuse it in more loaders.
- Several places in the code seem to use a mix of tabs and spaces; I'd suggest doing a global search and replace. Ideally CI would catch that.

**Comment by kylebarron** [comment]
> I included some utilities in deck.gl at [https://github.com/visgl/deck.gl/blob/master/modules/layers/src/geojson-layer/geojson-binary.js](geojson-binary.js) and I agree it could be useful. Should it be part of a new utils module? Or just a submodule of `loaders.gl`?

I thought `@loaders.gl/gis` was intended to have a range of binary utilities, but not sure what the best place is.

> For these to run and to reproduce the results stated above, you'll need to download some test data, see the script in `modules/mvt/test/data/fetch_data.py`.

Presumably this means the benchmarks will fail on the website https://loaders.gl/examples/benchmarks, where data won't be found.

**Comment by ibgreen** [comment]
> I thought @loaders.gl/gis was intended to have a range of binary utilities, but not sure what the best place is.

Yes. We also have similar utilities in math.gl, now with binary polygon support. If the code is sufficiently general, I would be open to have some of it there.

I have broken out the worker-utils in a separate module and ultimately I want to make that accessible from math.gl so we can offer `processOnWorker` compatible operations there.

Collecting things in the `@loaders.gl/gis` modules seems like a good starting point.

File: modules/mvt/src/lib/parse-mvt.js [hunk 19]
```diff
@@ -1,22 +1,37 @@
...
   * @param {arrayBuffer} _ A MVT arrayBuffer
-  * @return {?Object} A GeoJSON geometry object
+  * @return {?Object} A GeoJSON geometry object or a binary representation
   */
 export default function parseMVT(arrayBuffer, options) {
   options = normalizeOptions(options);
 
+  const binary = options.gis.format === 'binary';
   const features = [];
+  const firstPassData = {
```
Comment (Pessimistress) [inline modules/mvt/src/lib/parse-mvt.js:19]: One idea to generalize this would be something like a `BinaryFeatureWriter` class in `@loaders.gl/gis`:

```js
const writer = new BinaryFeatureWriter();

writer.addPoint(coordinates, properties);
writer.addLineString(coordinates, properties);
writer.addPolygon(coordinates, properties);

...

return writer.getData();
```

The class can allocate the typed arrays dynamically (arbitrary initial size; double size when overflown) without knowing how many features/vertices there will be.

This way, you don't have to modify the feature parser to get statistics, and you don't have to hold on to all features in memory.

File: modules/mvt/test/mvt-loader.binary.js (hunk 3 above)
Comment (felixpalmer) [reply modules/mvt/test/mvt-loader.binary.js:6]: This file was more for me to just sanity check my development, but I've now moved it into the spec test, as it doesn't hurt to keep it around. The `setLoaderOptions` is present there, but that seems to be the norm for *spec tests

File: modules/mvt/test/mvt-loader.bench.js (hunk 2 above)
Comment (felixpalmer) [reply modules/mvt/test/mvt-loader.bench.js:6]: Done

File: modules/mvt/test/data/fetch_data.py (hunk 4 above)
Comment (felixpalmer) [reply modules/mvt/test/data/fetch_data.py:1]: No this is not meant for CI, more to help me easily check that the new code is equivalent to the old. @kylebarron fixed

File: modules/mvt/src/mvt-loader.js (hunk 1 above)
Comment (felixpalmer) [reply modules/mvt/src/mvt-loader.js:25]: Good spot, this was indeed just testing. Will fix

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.d.ts (hunk 5 above)
Comment (felixpalmer) [reply modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.d.ts:5]: Added

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.d.ts (hunk 6 above)
Comment (felixpalmer) [reply modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.d.ts:3]: Done

File: modules/mvt/src/lib/binary-vector-tile/features-to-binary.js (hunk 7 above)
Comment (felixpalmer) [reply modules/mvt/src/lib/binary-vector-tile/features-to-binary.js:4]: Done

Comment (felixpalmer) [reply modules/mvt/src/lib/binary-vector-tile/features-to-binary.js:4]: @ibgreen OK so while I added the header, I cannot figure out what to do regarding the Typescript. If I remove the `ts- nocheck` directive then I get some errors (basically this issue https://stackoverflow.com/questions/52423842/what-is-not-assignable-to-parameter-of-type-never-error-in-typescript) and I don't see how I can fix them without converting the whole file to typescript (admittedly, I'm no Typescript expert). But I don't see any other place in the codebase that does this. Could you give some guidance on how to proceed?

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.js (hunk 8 above)
Comment (felixpalmer) [reply modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.js:1]: I would argue it makes sense to keep the classes together though, I feel that with the jsdocs it is now clear which files have the changes

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.d.ts (hunk 10 above)
Comment (felixpalmer) [reply modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.d.ts:15]: fixed

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.js (hunk 11 above)
Comment (felixpalmer) [reply modules/mvt/src/lib/binary-vector-tile/vector-tile-feature.js:200]: This, like a lot of the code is just what was in the original mapbox implementation, I think it is worth keeping it the same, as it makes it easier to reference against the original code

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js (hunk 12 above)
Comment (felixpalmer) [reply modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js:10]: Again, from the original implementation - I'm not clued up enough to comment

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js (hunk 13 above)
Comment (felixpalmer) [reply modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js:75]: Yes, I raised an eyebrow when I saw this too, but as per earlier comment, I would rather not change the original implementation if not necessary

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js (hunk 14 above)
Comment (felixpalmer) [reply modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js:50]: Again, original implementation

File: modules/mvt/src/lib/parse-mvt.js (hunk 15 above)
Comment (felixpalmer) [reply modules/mvt/src/lib/parse-mvt.js:109]: Done

File: modules/mvt/test/data/fetch_data.py (hunk 16 above)
Comment (felixpalmer) [reply modules/mvt/test/data/fetch_data.py:29]: The files chosen are large, so in total 31MB. I don't think it is worth storing them in the repo. However I think it is good to keep the script around in case we want to debug on some tiles that cause problems in the future & or we want to check performance against. As noted elsewhere, this script is not run as part of CI, that uses already checked in mvt test data to run sanity checks, which should guard against regressions

File: modules/mvt/test/data/fetch_data.py (hunk 17 above)
Comment (felixpalmer) [reply modules/mvt/test/data/fetch_data.py:32]: fixed

File: test/browser.js (hunk 18 above)
Comment (felixpalmer) [reply test/browser.js:38]: Er, not sure how I managed to do that - it was not intended

File: modules/mvt/src/lib/parse-mvt.js (hunk 19 above)
Comment (felixpalmer) [reply modules/mvt/src/lib/parse-mvt.js:19]: I think this could be interesting to explore in the future, but I would say out of scope for this PR.

**Comment by felixpalmer** [comment]
> Presumably this means the benchmarks will fail on the website https://loaders.gl/examples/benchmarks, where data won't be found.

@kylebarron I wasn't aware of that page, thanks for sharing. Perhaps it would be worth checking in some better benchmark example MVT tiles. The existing ones are so small the parsing speed improvement gets lost. Should I check in say, 3 tiles that demonstrate the speed improvement?

File: docs/whats-new.md [hunk 20]
```diff
@@ -43,6 +43,10 @@ Target Release Date: Q1 2021. (Alpha releases are available).
 
 - New table category loader for Excel spreadsheets in both binary `.xls`, `.xlsb` and XML-based `.xlsx` formats.
 
+**@loaders.gl/mvt**
+
+- Binary output is now 2-3X faster for large datasets thanks to parsing directly from PBF to binary, rather than going through geoJSON as an intermediate representation
```
Comment (kylebarron) [inline docs/whats-new.md:48]: Nit
```suggestion
- Binary output is now 2-3X faster for large datasets thanks to parsing directly from PBF to binary, rather than going through GeoJSON as an intermediate representation
```

**Review by kylebarron — APPROVED** [review]
I agree there's a benefit to not modifying code copied from the original source, though I'd love some comments or a way to know which code specifically is the way it is for that reason. Otherwise in a few months we might come back to this code and modify code unintentionally because we don't remember it was copied from upstream.

File: docs/whats-new.md (hunk 20 above)
Comment (alasarr) [reply docs/whats-new.md:48]: Would be great to include a table like this to explain the performance improvement:

https://github.com/visgl/deck.gl/blob/master/docs/whats-new.md#binary-mode

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js (hunk 12 above)
Comment (alasarr) [inline modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js:10]: Can we have this as a parameter with default to 4096? Some backend providers allow to specify the tile extent

Comment (alasarr) [reply modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js:10]: Forget about this, It would be a nice feature, but it should be addressed in a separated PR

**Review by alasarr — APPROVED** [review]

**Comment by alasarr** [comment]
Looks good to me. I think we might address some of the changes suggested and do the merge. 

After that, we could start working on support holes in binary mode (https://github.com/visgl/loaders.gl/issues/1274, and https://github.com/visgl/loaders.gl/issues/1128) and tesselation in workers (binary implementation of earcut at math.gl)

File: modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.d.ts [hunk 21]
```diff
@@ -5,7 +5,11 @@ import VectorTileFeature from './vector-tile-feature';
  * a vector tile layer. It has the same semantics
  * as the original library (https://github.com/mapbox/vector-tile-js)
  * except that the features it contains expose their data
- * in a binary optimized format, rather than geoJSON
+ * in a binary optimized format, rather than GeoJSON
+ *
+ * The VectorTileLayer class is unmodified from the original source,
+ * except that the `feature` method passes through an additional
+ * `firstPassData` parameter to the VectorTileFeature
```
Comment (felixpalmer) [inline modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.d.ts:12]: @kylebarron as requested, I've marked which code is different from the original Mapbox implemention
